libraries/standard/mqtt/src/mqtt.c

```diff
@@ -539,46 +628,88 @@ static MQTTStatus_t receivePacket( MQTTContext_t * const pContext,
 
 /*-----------------------------------------------------------*/
 
-static MQTTStatus_t sendPublishAcks( MQTTContext_t * const pContext,
-                                     uint16_t packetId,
-                                     MQTTPublishState_t * pPublishState,
-                                     bool stateNeedUpdate )
+static void getPacketTypeAndByte( MQTTPublishState_t publishState,
+                                  MQTTPubAckType_t * const pPacketType,
+                                  uint8_t * const pPacketTypeByte )
 {
-    MQTTStatus_t status = MQTTSuccess;
-    MQTTPublishState_t newState = MQTTStateNull;
-    int32_t bytesSent = 0;
-    uint8_t packetTypeByte = 0U;
-    MQTTPubAckType_t packetType;
-
-    assert( pContext != NULL );
-    assert( pPublishState != NULL );
-
-    switch( *pPublishState )
+    switch( publishState )
     {
         case MQTTPubAckSend:
-            packetTypeByte = MQTT_PACKET_TYPE_PUBACK;
-            packetType = MQTTPuback;
+            *pPacketTypeByte = MQTT_PACKET_TYPE_PUBACK;
+            *pPacketType = MQTTPuback;
             break;
 
         case MQTTPubRecSend:
-            packetTypeByte = MQTT_PACKET_TYPE_PUBREC;
-            packetType = MQTTPubrec;
+            *pPacketTypeByte = MQTT_PACKET_TYPE_PUBREC;
+            *pPacketType = MQTTPubrec;
             break;
 
         case MQTTPubRelSend:
-            packetTypeByte = MQTT_PACKET_TYPE_PUBREL;
-            packetType = MQTTPubrel;
+            *pPacketTypeByte = MQTT_PACKET_TYPE_PUBREL;
+            *pPacketType = MQTTPubrel;
             break;
 
         case MQTTPubCompSend:
-            packetTypeByte = MQTT_PACKET_TYPE_PUBCOMP;
-            packetType = MQTTPubcomp;
+            *pPacketTypeByte = MQTT_PACKET_TYPE_PUBCOMP;
+            *pPacketType = MQTTPubcomp;
             break;
 
         default:
             /* Take no action for states that do not require sending an ack. */
             break;
     }
+}
+
+/*-----------------------------------------------------------*/
+
+static MQTTStatus_t updateAckState( MQTTContext_t * const pContext,
+                                    uint16_t packetId,
+                                    bool stateNeedUpdate,
+                                    MQTTPubAckType_t packetType,
+                                    MQTTPublishState_t * const pNewState )
```

Comment:
This function does not seem necessary to me either, since the complexity is 8 if you put this code back in `sendPublishAcks` (at least due to the changes in #979). Also, the `pNewState` param will never be used by `sendPublishAcks` due to the fact that this functions returns `MQTTIllegalState` on error.